### DISCUSSION TITLE
I turned the serve-method into an APIView

### DISCUSSION THREAD

**Discussion by Tobeyforce** [wagtail/wagtail#8374] [General]
Could overriding the default serve-view in wagtail.core.views, attaching a django rest framework APIView decorator, cause major problems?

```python
wagtail.core.views
@api_view(['GET'])
def serve(request, path):
    # we need a valid Site object corresponding to this request in order to proceed
    site = Site.find_for_request(request)
    if not site:
        raise Http404

    path_components = [component for component in path.split('/') if component]
    page, args, kwargs = site.root_page.localized.specific.route(request, path_components)

    for fn in hooks.get_hooks('before_serve_page'):
        result = fn(page, request, args, kwargs)
        if isinstance(result, HttpResponse):
            return result

    return page.serve(request, *args, **kwargs)
```

Since I use wagtail as a headless CMS I ended up overriding the serve method on all page models demonstrated here, rather than using the wagtail API v2:
https://docs.wagtail.org/en/stable/reference/pages/model_recipes.html

The reason being that I prefer not using the PagesAPIViewset and instead going for a route of custom serializers attached to various page models. However, doing so meant I now had to use JsonResponse(), because the DRF Response() requires the view to be an APIView doing the content negotiation (as far as I understand) which will attach the correct renderer.

Any alternative solutions / ideas / feedback? One of the reasons I wanted to use Response() was simply to allow debug toolbar to work, because I couldn't manage to figure out a middleware allowing that using JsonResponse.  As of right now I really do feel like using wagtail headlessly is really fighting against my intuition a lot of times when working with pages.

**Comment by lb-** [comment]
Is this similar to the approach taken in this blog post?

https://dev.to/tommasoamici/headless-wagtail-what-are-the-pain-points-ji4

**Reply by Tobeyforce** [reply]
Yes, that's exactly right. This seems to be a common/dominant way of using wagtail headlessly, as demonstrated here: https://www.accordbox.com/blog/ebook-the-definitive-guide-to-nextjs-and-wagtail/
and here: https://github.com/Frojd/Wagtail-Pipit

Basically, overriding the get_context and serve method on the Page and using a serializer attached.

To be more specific, this is what my baseclass looks like for my pages.

```python
class BasePage(Page):
    serializer_class = None # Would normally add e.g "datafetcher.serializers.cms.RootPageSerializer"

    class Meta:
        abstract = True

    def get_component_data(self):
        # Custom method that takes the specified serializer defined and returns the pagename and data generated by the serializer to the serve method
        if not self.serializer_class:
            raise Exception(
                f'serializer_class is not set {self.__class__.__name__}')
        serializer_class = import_string(self.serializer_class)
        return {
            'page_type': self.__class__.__name__,
            'page_content': serializer_class(self).data
        }

    def get_context(self, request, *args, **kwargs):
        context = super().get_context(request)
        context['page_component'] = self.get_component_data()
        if 'root_page' not in context:
            context['root_page'] = RootPage.objects.first()
        return context
    
    def serve(self, request, *args, **kwargs):
        # Overrides the default serve behaviour that normally returns a template response, but we should return Json instead
        context = self.get_context(request, *args, **kwargs)
        return Response(data=context['page_component'], status=status.HTTP_200_OK)
```
For subclasses I'll also inherit from the RoutablePageMixin so that I can add more views / endpoints.

I do feel like I've been a bit torn, why not use the API v2 wagtail provides? It just seemed like my hands would be tied.

**Comment by kaedroho** [comment]
Thanks! I like this approach!

What do you think about making Wagtail look for a ``Accept: application/json`` header then make pages return an API response (using ``api_fields``) if this header is provided, but a HTML response if not?

We could also provide a ``serve_api`` method that could be overridden in a similar way that ``serve`` could be overridden to allow custom behaviour to be added (and we can leave ``serve`` as it is).

Also, I'm wondering why you had to add a custom serializer instead of using the PageSerializer in Wagtail?

**Reply by Tobeyforce** [reply]
I do think that there indeed should be something like a serve_api method and some kind of dynamic content negotiation the way DRF applies it in APIViews. The reason I never used the PageSerializer from wagtail was that I couldn't really wrap my head around how it worked, only looking at the source code.  Some documentation there would be greatly appreciated, it does seem very powerful to inherit from.

Ultimately, what I'm trying to allow for is having multiple pages, where each page can be accessed through different views that use different serializers that decide what data should be included. Since I'm very much new to wagtail I might be missing something, I do understand that wagtail seem to be very focused on the model definitions.

**Reply by kaedroho** [reply]
An idea to make something like this configurable: we could borrow Django REST Framework's [renderers](https://www.django-rest-framework.org/api-guide/renderers/) concept, where a "renderer" is a class that tells Wagtail how to render a page. Multiple renderers could be registered and one is selected based on the ``Accept`` header.

By default, we could provide one renderer, which just calls the ``serve`` method:

```python
WAGTAIL_PAGE_RENDERERS = [
    'wagtail.renderers.ServeMethod',
]
```

We could add to this an API renderer which will return an API response if the ``Accept: application/json`` header is provided. This could be specified on its own but can also be used with the ``ServeMethod`` renderer if you want to support both methods of rendering

```python
WAGTAIL_PAGE_RENDERERS = [
    'wagtail.renderers.ServeMethod',
    'wagtail.api.v2.renderers.APIV2Renderer',
]
```

Anyone could write their own renderer class if they want to customise the output.

Django REST Framework can also use the ``Accept`` header to specify a version, we could use this to allow developers to support two different API versions simultaneously:

```python
WAGTAIL_PAGE_RENDERERS = [
    'wagtail.api.v2.renderers.APIV2Renderer',
    'wagtail.api.v3.renderers.APIV3Renderer',
]
```

In this case, passing ``Accept: application/json; version=2`` would render the page with ``wagtail.api.v2.renderers.APIV2Renderer`` and  ``Accept: application/json; version=3`` would render the page with ``wagtail.api.v3.renderers.APIV3Renderer``.

I haven't thought much yet how this would work with ``RoutablePageMixin`` though. Is there anything else that this would need to work with?

What does everyone think of this?

**Reply by kaedroho** [reply]
I was writing the above response when you responded!

> The reason I never used the PageSerializer from wagtail was that I couldn't really wrap my head around how it worked, only looking at the source code. Some documentation there would be greatly appreciated, it does seem very powerful to inherit from.

I think this is mostly due to the way different Wagtail pages have different representations, but also the ``?fields`` GET parameter which allows developers to select a sub-set of fields on the page (useful for excluding body content if you don't need it, or selecting particular image renditions).

**Comment by INFINTY1999** [comment]
can someone help me out? how can I get help on the problems that I am facing.

**Reply by gasman** [reply]
Please see https://docs.wagtail.org/en/stable/support.html

**Comment by mdfathi99** [comment]
I think this might help for JSONRender.

#api.py
```
from rest_framework.renderers import JSONRenderer

...

class ProdPagesAPIViewSet(PagesAPIViewSet):
    renderer_classes = [JSONRenderer]
    name = "pages"

api_router.register_endpoint("pages", ProdPagesAPIViewSet)
```

**Comment by lb-** [comment]
We have now raised an issue to try to define this pattern https://github.com/wagtail/wagtail/issues/11752

If someone is up for a pretty sizable contribution to Wagtail, it may be a good chance to get involved.

This issue builds on the work done in https://github.com/wagtail/wagtail/pull/8386 and tries to define what other items would need to be addressed before we could look at this as being merged into the core code.
